cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -1565,5 +1589,43 @@ pub fn define(
     e.enc_both(ffcmp.bind(F32), rec_fcmp.opcodes(vec![0x0f, 0x2e]));
     e.enc_both(ffcmp.bind(F64), rec_fcmp.opcodes(vec![0x66, 0x0f, 0x2e]));
 
+    // SIMD splat: before x86 can use vector data, it must be moved to XMM registers; see legalize.rs for
+    // how bitcast is used to move the data there
+
+    let mut shuffle_mapping: HashMap<u64, Vec<u8>> = HashMap::new();
+    shuffle_mapping.insert(8, vec![0x66, 0x0f, 0x38, 00]); // PSHUFB from SSSE3, TODO needs another instruction to get the mask into another register (plus bitcast)
+    // TODO legalize 16x8 to two PINSRW (from SSE), then PSHUFD as if it were a 32x4?
+    shuffle_mapping.insert(32, vec![0x66, 0x0f, 0x70]); // PSHUFD from SSE2
+    // TODO legalize 64x2 as two PINSRQ instructions (from SSE 4.1)?
+
+    for ty in LaneTypeIterator::new() {
+        if let Some(opcode) = shuffle_mapping.get(&ty.lane_bits()) {
+            let number_of_lanes = 128 / ty.lane_bits();
+            e.enc32(x86_pshuf.bind_vector(ty, number_of_lanes), rec_r_ib_unsigned.nonrex().opcodes(opcode.clone()));
+            e.enc64(x86_pshuf.bind_vector(ty, number_of_lanes), rec_r_ib_unsigned.nonrex().opcodes(opcode.clone()));
+            // TODO can't use enc_both here because it forces a .rex() encoding
```

Comment:
This is the block that performs the `PSHUF` encodings; see the TODO comments for the missing parts or parts that need to be legalized. One question here: when I add a settings predication (e.g. with `enc64_isap`) and then attempt to write a `binemit` test targeting that feature set (e.g. `set use_sse41`), cranelift cannot seem to find an encoding--is there something I'm missing here? Once I figure that out I can add the settings predicates back in.

Reply:
Currently, tests handle this by selecting an explicit CPU, like this: `target x86_64 haswell`. That should hopefully unblock you for now, though ideally we should add a better way to handle architecture flags like `use_sse41`.

Reply:
But I see code like `set is_pic` in `filetests`; how come `is_pic` can be modified but I can't modify the other settings from the tests?

Reply:
Ah, I forgot that you can set a flag on the target line too: `target x86_64 use_sse41=true`. That's the proper way to do this.